networkx/algorithms/approximation/traveling_salesman.py

```diff
@@ -321,6 +326,368 @@ def threshold_tsp(G, weight):
+    ----------
+    .. [1] A. Asadpour, M. X. Goemans, A. Madry, S. O. Gharan, and A. Saberi,
+       An o(log n/log log n)-approximation algorithm for the asymmetric
+       traveling salesman problem, Operations research, 65 (2017),
+       pp. 1043–1061
+
+    .. [2] M. Held, R. M. Karp, The traveling-salesman problem and minimum
+           spanning trees, Operations Research, 1970-11-01, Vol. 18 (6),
+           pp.1138-1162
+    """
+
+    def k_pi():
+        """
+        Find the set of minimum 1-Arborescences for G at point pi.
+
+        Returns
+        -------
+        Set
+            The set of minimum 1-Trees
+        """
+        # Create a copy of G without vertex 1
+        G_1 = G.copy()
+        # node is node '1' in the Held and Karp paper
+        G_1.remove_node(node)
+        minimum_arborescence_weight = math.inf
+        minimum_1_arborescence_weight = math.inf
+        minimum_1_arborescences = set()
+        for arborescence in nx.ArborescenceIterator(G_1):
+            if minimum_arborescence_weight == math.inf:
+                minimum_arborescence_weight = arborescence.size(weight)
+            elif arborescence.size(weight) > minimum_arborescence_weight:
+                # We are done with minimum arborescences
+                break
+            arborescence.add_node(node)
+            # We can pick the minimum weight in-edge for the vertex with a cycle
+            min_in_edge_weight = math.inf
+            min_in_edge = None
+            for u, v, d in G.in_edges(node, data=True):
+                edge_weight = d[weight]
+                if edge_weight < min_in_edge_weight:
+                    min_in_edge_weight = edge_weight
+                    min_in_edge = (u, v)
+            arborescence.add_edge(
+                min_in_edge[0], min_in_edge[1], weight=min_in_edge_weight
+            )
+            # We have to pick the root node of the arborescence for the out edge
+            # of the first vertex as that is the only node without an edge
+            # directed into it.
+            for n in arborescence:
+                if arborescence.in_degree(n) == 0:
+                    # root found
+                    arborescence.add_edge(node, n, weight=G[node][n][weight])
+                    break
+            # We now have a 1-arborescence
+            if arborescence.size(weight) < minimum_1_arborescence_weight:
+                minimum_1_arborescence_weight = arborescence.size(weight)
+            minimum_1_arborescences.add(arborescence)
+        # Not every minimum spanning arborescence will produce a minimum
+        # 1-arborescence, depending on the cheapest root to connect to.
+        # Once we have this set, we need to delete all of the non-minimum
+        # 1-arborescences from it.
+        for one_arborescence in list(minimum_1_arborescences):
+            if one_arborescence.size(weight) > minimum_1_arborescence_weight:
+                minimum_1_arborescences.discard(one_arborescence)
```

Comment:
Can you put the min weight check into the loop, so you don't have to hold everything in memory (I don't know how big `minimum_1_arborescences` can get).  One way to do that is to switch from a set to a dict keyed by arborescence to the arb.size(weight) value.  That saves recomputing the size.  Then something like:
 ```suggestion
            new_weight = arborescence.size(weight)
            for other_arb, arb_weight in list(minimum_1_arborescences.items()):
                if arb_weight > new_weight:
                    del minimum_1_arborescence[other_arb]
            minimum_1_arborescence[arborescence] = new_weight
```

Reply:
`minimum_1_arborescences` is generally pretty small. For my test that I'm currently working with, it is normally only one or two arborescences but I have seen it get as big as four. 

Take a look at the changes I pushed up on June 21st, I completely rewrote that method in a (failed) attempt to find the optimal solution by finding the minimum arborescences while changing which node creates the cycle. This will cause the program to oscillate between two different arborescences and never terminate but the new code should be smoother.